### ISSUE TITLE
Review CachedReflectionInfo class to validate method overloades and unused types

### ISSUE THREAD

**Issue by adityapatwardhan** [PowerShell/PowerShell#9016]
Please review the code under [Compiler.cs](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs) and look at all the methods we cache. We need to specify the parameters too when we do `GetMethod` on a type so we explicitly load a particular overload.

See #9015 

# Environment data

```none
Name                           Value
----                           -----
PSVersion                      6.1.3
PSEdition                      Core
GitCommitId                    6.1.3
OS                             Microsoft Windows 10.0.18845
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

```

**Comment by iSazonov** [comment]
Interesting, can we automate this?

**Comment by iSazonov** [comment]
@daxian-dbw Make sense add types for parameters in `GetMethod`? It is a lot of work...

**Comment by daxian-dbw** [comment]
We found pwsh crashing on the latest build of .NET Core 3.0 and it turned out that's because we have code like `typeof(String).GetMehtod("get_Chars")` which assumes the public method/property has no overloads of 'Chars' in `String` type, however, 2 more overloads of `Chars` get added to `String` in .NET Core 3.

So for code like `GetMethod(string)`, `GetProperty(string)` and similar call sites, we need to make it explicit by specifying the parameter types, so we won't get into the same pit again in future.

**Comment by iSazonov** [comment]
@daxian-dbw I understand where the issue come from. My question is, can we select a subset that needs to be corrected or should we correct _all_ the lines, since we cannot predict which api will be expanded with a new overload?

**Comment by daxian-dbw** [comment]
I get your point. Maybe it isn't worth the effort. Identifying the root cause of issues like this is not trivial, but is not super hard too.
I would like to, **gradually**, update the instances in `CachedReflectionInfo`, which work with public instance or static members of public types, to use the overload of `GetMethod/GetProperty` that require the parameter type argument. It's by no means an urgent work.

But I want us to be more future proof on this, meaning that when reviewing code, spot `GetMethod(string)` or similar that are called on public types, and use the more explicit overload instead.

**Comment by daxian-dbw** [comment]
Fixed via #9029 and #9019
